### ISSUE TITLE
Flash.now is not rendered when form_with is used

### ISSUE THREAD

**Issue by mdorr** [rails/rails#29036]
### Steps to reproduce

Consider this controller action:

```ruby
def create
  @user = User.new(user_params)
  if @user.save
    flash[:notice] = "User saved to database."
    redirect_to user_url @user
  else
    flash.now[:alert] = "Error: #{@user.errors.full_messages.join(', ')}"
    render :new
  end
end
```

This model:
```ruby
class User < ApplicationRecord
  validates :name, presence: true
end
```

And this form:
```ruby
<%= form_with @user do |f| %>
  <%= f.label :name %>
  <%= f.text_field :name %>
  <%= f.submit %>
<% end %>
```

This is pretty standard. Elsewhere in your application, the flash would be rendered if it is set.

The issue appears when we try to enter an empty name. Validation fails, and we should see a flash with the error message.

However, if the form that is used to call this is using the new `form_with` method, flash.now is not set. If you enter invalid input and cause the `else` path to be tripped, the flash will never be set. Using the older `form_for` method works as expected

### Expected behavior

flash.now should be set and be displayed.

### Actual behavior

The flash is not set. Including `flash debug` in the page shows that it is not set. Additionally, the page does not reload - meaning it does not jump back to the top of the page. This might indicate that `form_with` does some caching, which in this case causes the issue.

Using the old `form_for` helper fixes the issue, which leads me to believe the issue is caused by the new `form_with` helper.

### System configuration
**Rails version**: 5.1.0

**Ruby version**: ruby 2.4.0p0 (2016-12-24 revision 57164) [x86_64-darwin15]

**Google Chrome**: 
- Version: 57.0.2987.133 (Official Build) (64-bit) 
- Revision: ec33cd0c06881d919ac0de419d829ad914e0be8f-refs/branch-heads/2987@{#887}
- OS: Mac OS X 
- JavaScript: V8 5.7.492.71

**Comment by morgoth** [comment]
Can you try with `<%= form_with @user, local: true do |f| %>`, as the docs says `By default form submits are remote and unobtrusive XHRs.` which results in problems displaying errors.

Unfortunately, I think it will be a common issue for all. @kaspth Did you think about this?

**Comment by mdorr** [comment]
Thank you, `local: true` fixed the issue - I missed that in the docs. This is different default behavior from the other form helpers though, isn't it? That will be confusing to some. Is there a benefit in switching this for `form_with` specifically?

**Comment by sunnyrjuneja** [comment]
Hi, I ran into this problem today. I think it makes sense to update flash to respond XHR. Thoughts @maclover7 ?

**Comment by cjmao** [comment]
Also encountered today!
Using a `form_with` helper (without setting `local: true`) in the view, setting a `flash.now` (followed by a `render 'new'`) in the controller, and then submitting an empty form in the browser will cause this.

**Comment by MSCAU** [comment]
I agree that UJS / remote is great, but the way flash works, it seems to rely on the controller responding with HTML rather than JS. So currently, if you use UJS, you can't get the benefit of flash functionality. Or am I missing something?

**Comment by wlycdgrfromflatiron** [comment]
Banged my head against this for two hours today before realizing - thanks to this post - that this is why render :new from #create (on resource validation error, you know) wasn't updating the view in browser as expected.

**Comment by npras** [comment]
Is setting `local: true` the only option still as @morgoth suggested?

This means we'd have to make the form "non-ajaxy" just for the `else` part of the controller action where we'd render instead of redirect.

With ajax submit (and with a wrong form input, eg: invalid password), I can see in the devtools network tab the html response from the server (in the Response preview section). It has the flash message rendered properly. I wonder how the javascript receiving this html can be forced to replace what's there already.

Should we use Turbolink's clearcache() etc? I tried though, but didn't work, although I might not have tried correctly.

**Comment by 7uthaifah** [comment]
> Thank you, `local: true` fixed the issue - I missed that in the docs. This is different default behavior from the other form helpers though, isn't it? That will be confusing to some. Is there a benefit in switching this for `form_with` specifically?

Simple, worked for me

**Comment by vvo** [comment]
Adding to this issue: flash messages seems to be the idiomatic way to inform the users that something happened but they are not displayed when using the idiomatic way to build forms in recent Rails version (form_with). This is very confusing for new users like me I wish there was a way to catch flash messages in JS in a generic way but can't find it.

**Comment by sprite2005** [comment]
Had this same issue. Couldn't figure out why my flash message wasn't displaying. Luckily this issue helped me resolve it.

**Comment by anhari** [comment]
If you've reached this far, you might be interested in this issue: https://github.com/turbolinks/turbolinks-rails/issues/40.

It offers the [turbolinks_render](https://github.com/jorgemanrubia/turbolinks_render) gem as a solution, and hints that this may work by default later on down the road.

**Comment by jedharwood** [comment]
I have a related issue;
I was returning;
```
else
    flash.now[:error] = "Error: City could not be saved"
    render :new, locals: { city: @city }
end
```
And that was rendering the city/new view and pre-populating it with the information already entered by the user. 

I also noticed that I didn't need to use `locals: {}` and that I could just `render :new` and that would also pre-populate my form. Thanks, Rails.

It would not, however, display the flash notice. I updated the code to;
`render :new, status: :unprocessable_entity` 
and that solved the issue of the flash notice not displaying. But now it won't pre-populate the form. Is there a way I can have the `:new` page render with the form pre-populated with the user entered values and also show the flash message?
